src/VisualStudio/Core/Def/Implementation/Snippets/AbstractSnippetExpansionClient.cs

```diff
@@ -393,7 +451,338 @@ public virtual bool TryInsertExpansion(int startPositionInSubjectBuffer, int end
+                                    new XElement(
+                                        snippetNamespace + "Literal",
+                                        new XAttribute(snippetNamespace + "Editable", "true"),
+                                        new XElement(snippetNamespace + "ID", new XText("placeholder")),
+                                        new XElement(snippetNamespace + "ToolTip", new XText("")),
+                                        new XElement(snippetNamespace + "Default", new XText("")),
+                                        new XElement(snippetNamespace + "Function"))),
+                                new XElement(
+                                    snippetNamespace + "Code",
+                                    new XAttribute(snippetNamespace + "Language", "csharp"),
+                                    new XCData(template))))));
+
+                var doc = new DOMDocumentClass();
+                if (doc.loadXML(snippet.ToString(SaveOptions.OmitDuplicateNamespaces)))
+                {
+                    var methodSymbols = symbols.OfType<IMethodSymbol>().ToImmutableArray();
+                    _symbols = methodSymbols;
+                    _symbol = null;
+
+                    if (expansion.InsertSpecificExpansion(doc, textSpan, this, LanguageServiceGuid, pszRelativePath: null, out ExpansionSession) == VSConstants.S_OK)
+                    {
+                        Debug.Assert(_symbols == methodSymbols);
+                        Debug.Assert(_symbol == null);
+
+                        // Trigger signature help after starting the snippet session
+                        var workspace = (VisualStudioWorkspace)SubjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges()!.Project.Solution.Workspace;
+                        var exportProvider = (IMefHostExportProvider)workspace.Services.HostServices;
+                        var editorCommandHandlerServiceFactory = exportProvider.GetExports<IEditorCommandHandlerServiceFactory>().Single().Value;
+                        var editorCommandHandlerService = editorCommandHandlerServiceFactory.GetService(TextView, SubjectBuffer);
+                        editorCommandHandlerService.Execute((view, buffer) => new InvokeSignatureHelpCommandArgs(view, buffer), nextCommandHandler: null);
+
+                        ThreadingContext.JoinableTaskFactory.RunAsync(async () =>
+                        {
+                            await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(CancellationToken.None);
+                            while (ExpansionSession is not null)
+                            {
+                                var controller = Controller.TryGetInstance(TextView, SubjectBuffer);
+                                if (controller is null)
+                                {
+                                    await Task.Yield();
+                                    continue;
+                                }
```

Comment:
uhhhhhhhhhh.  scary beyond belief.

Reply:
tagging @jasonmalinowski there's got to be a good pattern to accomplish this.  for example, SigHelp and Completion communicate with each other (IIRC), so having snippet communicate with sighelp (or vice versa) in a cleaner fashion seems like it must be doable.

Reply:
Documented a TODO linking to #50037

Reply:
So I guess I have some specific concerns here:

1. This is basically spinning a CPU thread until the other threads complete signature help. The contention won't make it come any faster. ("Are we there yet? Are we there yet?")
2. If anything else means the signature help doesn't trigger for _any_ reason, the UI locks up.
3. Whether the work being scheduled by signature help will actually be scheduled at a high enough to beat the Task.Yield. If it's not, then we're just going to hang.

Reply:
📝 This is now fixed by extracting a common `ControllerProvider` MEF class